apps/prairielearn/src/question-servers/freeform.ts

```diff
@@ -2080,20 +2084,46 @@ async function getContext(question: Question, course: Course): Promise<QuestionP
 
 async function getCacheKey(
   course: Course,
+  variant: Variant,
+  submission: Submission | null,
   data: ExecutionData,
   context: QuestionProcessingContext,
-) {
+): Promise<string | null> {
   try {
     const commitHash = await getOrUpdateCourseCommitHash(course);
-    const dataHash = objectHash({ data, context }, { algorithm: 'sha1', encoding: 'base64' });
-    return `question:${commitHash}-${dataHash}`;
+
+    const dataHash = objectHash(
+      [
+        // We deliberately exclude large user-controlled objects from the cache key.
+        // Whenever these change, the `modified_at` column of `variants` and/or
+        // `submissions` will change, which will cause the cache to be invalidated.
+        _.omit(data, [
+          'params',
+          'correct_answers',
+          'submitted_answers',
+          'format_errors',
+          'partial_scores',
+          'raw_submitted_answers',
+        ]),
+        context,
+        variant.modified_at,
+        submission?.modified_at,
+      ],
+      { algorithm: 'sha1', encoding: 'base64' },
+    );
+
+    // The variant and submission IDs are included in the cache key to ensure
+    // that data never leaks between variants or submissions.
+    return `variant:${variant.id}:submission:${submission?.id}:${commitHash}-${dataHash}:cache`;
```

Comment:
This is new. Our hashes are in all likelihood totally fine, but for the price of a few extra bytes in the key, we can be really sure that we're not accidentally getting cache hits when we shouldn't.

Reply:
I'm wondering if we should include the phase here? The cache is shared between `render` and `file`. The different bits of data present in each mean we should basically always end up with different cache keys already.

Reply:
More or less, yes: the commit hash is always updated as part of the syncing process.

It's not strictly atomic with the availability of new files on disk, but that's pretty difficult to achieve and we haven't considered this to be anything approaching a major problem.

Reply:
Makes sense, thanks.